python/ray/tune/function_runner.py

```diff
@@ -41,105 +47,184 @@ def __call__(self, **kwargs):
         Example:
             >>> reporter(mean_accuracy=1, training_iteration=4)
             >>> reporter(mean_accuracy=1, training_iteration=4, done=True)
+
+        Raises:
+            StopIteration: A StopIteration exception is raised if the trial has
+                been signaled to stop.
         """
+        
+        assert self._last_report_time is not None, (
+                "StatusReporter._start() must be called before the first "
+                "report __call__ is made to ensure correct runtime metrics."
+                )
 
-        with self._lock:
-            self._latest_result = self._last_result = kwargs.copy()
-        self._iteration += 1
-
-    def _get_and_clear_status(self):
-        if self._error:
-            raise TuneError("Error running trial: " + str(self._error))
-        if self._done and not self._latest_result:
-            if not self._last_result:
-                raise TuneError("Trial finished without reporting result!")
-            logger.warning("Trial detected as completed; re-reporting "
-                           "last result. To avoid this, include done=True "
-                           "upon the last reporter call.")
-            self._last_result.update(done=True)
-            self._last_result.setdefault(TRAINING_ITERATION, self._iteration)
-            return self._last_result
-        with self._lock:
-            res = self._latest_result
-            self._latest_result = None
-            if res:
-                res.setdefault(TRAINING_ITERATION, self._iteration)
-            return res
+        # time per iteration is recorded directly in the reporter to ensure
+        # any delays in logging results aren't counted
```

Comment:
can we do `kwargs.copy()` just in case?

Reply:
(sorry, should have caught this in an earlier review)

Reply:
I'm fairly certain python create a new dict at every call, but, given the cost of copying should be low and somewhat infrequent, I've added it for piece of mind.